src/data_structures/fmindex.rs

```diff
@@ -88,41 +100,78 @@ pub trait FMIndexable {
 
 }
 
-/// The Fast Index in Minute space (FM-Index, Ferragina and Manzini, 2000) for finding suffix array
-/// intervals matching a given pattern.
+pub trait FMIndexCore {
+    type SA: SuffixArray;
 
-#[cfg_attr(feature = "serde_macros", derive(Serialize, Deserialize))]
-pub struct FMIndex<
-        DBWT: DerefBWT + Clone,
-        DLess: DerefLess + Clone,
-        DOcc: DerefOcc + Clone> {
-    bwt: DBWT,
-    less: DLess,
-    occ: DOcc,
+    fn occ(&self, r: usize, a: u8) -> usize;
+    fn less(&self, a: u8) -> usize;
+    fn bwt(&self) -> &BWT;
+    fn sa(&self) -> &Self::SA;
 }
 
-impl<
-    DBWT: DerefBWT + Clone,
-    DLess: DerefLess + Clone,
-    DOcc: DerefOcc + Clone> FMIndexable for FMIndex<DBWT, DLess, DOcc> {
+impl FMIndexCore for SampledSuffixArray {
+    // the trait needs to track what kind of suffix array reference it's returning
+    type SA = SampledSuffixArray;
```

Comment:
Exactly. We definitely want the FMIndex to be able to live without the SA. A good example why this is useful is the BWA read aligner. "bwa aln" creates suffix array intervals, "bwa samse/sampe" converts these into genome positions. Also, it is possible to find out about uniqueness of substrings by just looking at the size of suffix array intervals.